### PR TITLE
Chore/retrieval chain to lcel

### PR DESCRIPTION
**Pull request by bracesproul**
Updating retrieval QA chain docs to use LCEL
Added export of `QA_PROMPT_SELECTOR`

TODO
- [x] update base example
- [x] update Custom QA example
- [x] update Custom Docs example
- [x] update Return Source Documents example
- [x] Hide legacy non runnable docs w/ link to find

### CODE REVIEW

File: examples/src/chains/retrieval_qa_custom.ts [hunk 1]
```diff
@@ -1,32 +1,86 @@
...
+
+const query = "What did the president say about Justice Breyer?";
+
+// Create a vector store retriever from the documents.
+const vector_store_retriever = (
+  await HNSWLib.fromDocuments(docs, new OpenAIEmbeddings())
+).asRetriever();
+const relevant_docs = await vector_store_retriever.getRelevantDocuments(query);
+
+const serialized_docs_content = relevant_docs.map((doc) => doc.pageContent);
```
Comment (jacoblee93) [inline examples/src/chains/retrieval_qa_custom.ts:56]: Prefer camel case

File: examples/src/chains/retrieval_qa_custom.ts [hunk 2]
```diff
@@ -1,32 +1,86 @@
...
+
+// Create a new chain using the default prompt and an output parser.
+const combine_docs_chain = combine_docs_prompt
+  .pipe(model)
+  .pipe(new StringOutputParser());
+
+const query = "What did the president say about Justice Breyer?";
+
+// Create a vector store retriever from the documents.
+const vector_store_retriever = (
```
Comment (jacoblee93) [inline examples/src/chains/retrieval_qa_custom.ts:51]: We generally try to break declarations up like this in examples:

```typescript
const vectorstore = await HNSWLib.fromDocuments()
const retriever = vectorstore.asRetriever();
```

File: examples/src/chains/retrieval_qa_sources.ts [hunk 3]
```diff
@@ -4,27 +4,74 @@ import { HNSWLib } from "langchain/vectorstores/hnswlib";
...
+   * The query to perform against the documents.
+   */
+  query: string;
+  /**
+   * The text to be chunked and used as context.
+   */
+  text: string;
+}): Promise<{ result: string; sourceDocuments: Array<Document> }> {
+  // Initialize the LLM to use to answer the question.
+  const model = new OpenAI({});
```
Comment (jacoblee93) [inline examples/src/chains/retrieval_qa_sources.ts:32]: Use `ChatOpenAI` for the later GPT-3.5 models or use `OpenAI` with `gpt-3.5-turbo-instruct`

File: langchain/src/chains/index.ts [hunk 4]
```diff
@@ -99,3 +99,4 @@ export {
   type OpenAPIChainOptions,
   createOpenAPIChain,
 } from "./openai_functions/openapi.js";
+export { QA_PROMPT_SELECTOR } from "./question_answering/stuff_prompts.js";
```
Comment (jacoblee93) [inline langchain/src/chains/index.ts:102]: See above, remove

**Comment by jacoblee93** [comment]
I think we want to move these into separate examples/.mdx files but broadly looks good!

File: examples/src/chains/retrieval_qa.ts [hunk 5]
```diff
@@ -1,33 +1,58 @@
...
 import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
 import * as fs from "fs";
+import {
+  RunnablePassthrough,
+  RunnableSequence,
+} from "langchain/schema/runnable";
+import { Document } from "langchain/document";
+import { PromptTemplate } from "langchain/prompts";
+import { StringOutputParser } from "langchain/schema/output_parser";
+import { QA_PROMPT_SELECTOR } from "langchain/chains";
```
[the remark this answers is not in the document]
Comment (bracesproul) [reply examples/src/chains/retrieval_qa.ts]: sg, removed export and replaced with inline string prompt

File: examples/src/chains/retrieval_qa_sources.ts (hunk 3 above)
Comment (bracesproul) [reply examples/src/chains/retrieval_qa_sources.ts]: Refactored to `ChatOpenAI` and use `ChatPromptTemplate` w/ system & human prompt (did the same for other examples except `retrieval_qa_custom`)

File: examples/src/chains/retrieval_qa_custom.ts [hunk 6]
```diff
@@ -1,32 +1,86 @@
...
+// Create a vector store retriever from the documents.
 const vectorStore = await HNSWLib.fromDocuments(docs, new OpenAIEmbeddings());
+const retriever = vectorStore.asRetriever();
+
+const relevantDocs = await retriever.getRelevantDocuments(query);
+
+const serializedDocsContent = relevantDocs.map((doc) => doc.pageContent);
+
+// Perform a map over all the relevant docs and query the LLM.
+const queryDocsChainResults = await Promise.all(
```
Comment (jacoblee93) [inline examples/src/chains/retrieval_qa_custom.ts:59]: Use a `mapReduce` chain here directly as discussed

Comment (bracesproul) [reply examples/src/chains/retrieval_qa_custom.ts]: Pushed up a commit updating it to use `loadQAMapReduceChain`, however I couldn't see where I could also use Runnables. Could this just be an instance where Runnables aren't ready for that type of implementation?

**Comment by jacoblee93** [comment]
Thank you!

**Pull request merged**